### PR TITLE
All requests now require that the user create a request object

### PR DESCRIPTION
**Pull request by ColdenCullen**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/request_response.c [hunk 1]
```diff
@@ -220,6 +220,34 @@ int aws_http_request_add_header(struct aws_http_request *request, struct aws_htt
     return AWS_OP_ERR;
 }
 
+int aws_http_request_add_header_array(
```
Comment (justinboswell) [inline source/request_response.c]: Thank you for this.

Comment (ColdenCullen) [reply source/request_response.c]: :+1:

File: source/h1_connection.c [hunk 2]
```diff
@@ -677,45 +696,55 @@ static void s_stream_write_outgoing_data(struct h1_stream *stream, struct aws_io
...
+                    /* Return success because we want to try again later */
+                    return AWS_OP_SUCCESS;
                 }
 
-                size_t prev_len = dst->len;
-
-                enum aws_http_outgoing_body_state state =
-                    stream->base.stream_outgoing_body(&stream->base, dst, stream->base.user_data);
+                struct aws_stream_status status;
+                aws_input_stream_get_status(stream->base.outgoing_body, &status);
```
Comment (graebm) [inline source/h1_connection.c]: check return value
`int err = aws_input_stream_get_status(...);`

File: source/h1_connection.c [hunk 3]
```diff
@@ -677,45 +696,55 @@ static void s_stream_write_outgoing_data(struct h1_stream *stream, struct aws_io
...
+                if (!status.is_valid) {
+                    AWS_LOGF_ERROR(
+                        AWS_LS_HTTP_STREAM, "id=%p: Body stream reporting status invalid", (void *)&stream->base);
+                    return AWS_OP_ERR;
```
Comment (graebm) [inline source/h1_connection.c]: Need to raise an error if aws_input_stream_get_status() succeeds, but `status.is_valid` is false

File: source/h1_connection.c [hunk 4]
```diff
@@ -677,45 +696,55 @@ static void s_stream_write_outgoing_data(struct h1_stream *stream, struct aws_io
...
+                }
 
-                AWS_FATAL_ASSERT(prev_len <= dst->len);
+                size_t amount_read = 0;
+                if (aws_input_stream_read(stream->base.outgoing_body, dst, &amount_read)) {
```
Comment (graebm) [inline source/h1_connection.c]: ok, this is the last nitpick comment I'll make about `err = aws_xyx(); if (err)`

File: source/h1_connection.c [hunk 5]
```diff
@@ -677,45 +696,55 @@ static void s_stream_write_outgoing_data(struct h1_stream *stream, struct aws_io
...
 
                 AWS_LOGF_TRACE(
                     AWS_LS_HTTP_STREAM,
                     "id=%p: Writing %zu body bytes to message.",
                     (void *)&stream->base,
-                    dst->len - prev_len);
+                    amount_read);
 
-                if (state == AWS_HTTP_OUTGOING_BODY_DONE) {
+                aws_input_stream_get_status(stream->base.outgoing_body, &status);
```
Comment (graebm) [inline source/h1_connection.c]: check return code

File: source/request_response.c [hunk 6]
```diff
@@ -220,6 +220,34 @@ int aws_http_request_add_header(struct aws_http_request *request, struct aws_htt
...
+        if (aws_http_request_add_header(request, headers[i])) {
+            goto error;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+
+error:
+    /* Remove all headers we added */
+    for (size_t len = aws_array_list_length(&request->headers); len > beginning_headers_size; --len) {
```
Comment (graebm) [inline source/request_response.c]: ooh, nice catch on the esoteric error handling ;)

File: source/request_response.c [hunk 7]
```diff
@@ -220,6 +220,34 @@ int aws_http_request_add_header(struct aws_http_request *request, struct aws_htt
...
 }
 
+int aws_http_request_add_header_array(
+    struct aws_http_request *request,
+    struct aws_http_header *headers,
+    size_t num_headers) {
+
+    AWS_PRECONDITION(request);
+    AWS_PRECONDITION(headers);
+    AWS_PRECONDITION(num_headers > 0);
```
Comment (graebm) [inline source/request_response.c]: trivial/debatable/whocares: We could allow num_headers == 0

File: source/websocket_bootstrap.c [hunk 8]
```diff
@@ -117,13 +106,23 @@ int aws_websocket_client_connect(const struct aws_websocket_client_connection_op
...
+    if (!options->allocator || !options->bootstrap || !options->socket_options || !options->host.len || !path.len ||
         !options->on_connection_setup) {
 
         AWS_LOGF_ERROR(AWS_LS_HTTP_WEBSOCKET_SETUP, "id=static: Missing required websocket connection options.");
         return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
     }
 
+    struct aws_byte_cursor method;
+    aws_http_request_get_method(options->handshake_request, &method);
+    if (!aws_byte_cursor_eq(&method, &s_method_get)) {
```
Comment (graebm) [inline source/websocket_bootstrap.c]: We already have a string->enum function

`if (aws_http_str_to_method(method) != AWS_HTTP_METHOD_GET)`

but we only have a HEAD enum right now, you'll need to create GET

File: source/websocket_bootstrap.c [hunk 9]
```diff
@@ -138,10 +137,10 @@ int aws_websocket_client_connect(const struct aws_websocket_client_connection_op
         return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
     }
 
-    if (!options->handshake_header_array || !options->num_handshake_headers) {
+    if (!options->handshake_request) {
         AWS_LOGF_ERROR(
             AWS_LS_HTTP_WEBSOCKET_SETUP,
-            "id=static: Invalid connection options, missing required headers for websocket client handshake.");
+            "id=static: Invalid connection options, missing required request websocket client handshake.");
```
Comment (graebm) [inline source/websocket_bootstrap.c]: "required request **for** websocket client handshake"

File: source/request_response.c (hunk 6 above)
Comment (ColdenCullen) [reply source/request_response.c]: :+1:

File: tests/test_tls.c [hunk 10]
```diff
@@ -182,15 +182,19 @@ static int s_test_tls_download_medium_file(struct aws_allocator *allocator, void
...
-        {.name = aws_byte_cursor_from_c_str("Host"), .value = *aws_uri_host_name(&uri)}};
+    struct aws_http_request *request = aws_http_request_new(allocator);
+    aws_http_request_set_method(request, aws_byte_cursor_from_c_str("GET"));
+    aws_http_request_set_path(request, *aws_uri_path_and_query(&uri));
+
+    struct aws_http_header header_host = {
+        .name = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Host"),
+        .value = *aws_uri_host_name(&uri),
+    };
+    aws_http_request_add_header(request, header_host);
```
Comment (graebm) [inline tests/test_tls.c]: ASSERT_SUCCESS() around the set_method/set_path/add_header calls

File: source/h1_connection.c [hunk 11]
```diff
@@ -284,14 +286,19 @@ static bool s_connection_is_open(const struct aws_http_connection *connection_ba
...
     size_t total = 0;
 
+    const size_t num_headers = aws_http_request_get_header_count(request);
+
     for (size_t i = 0; i < num_headers; ++i) {
-        struct aws_http_header header = header_array[i];
+        struct aws_http_header header;
+        if (aws_http_request_get_header(request, &header, i)) {
+            AWS_FATAL_ASSERT(false); /* i is guaranteed to be within the valid index range */
+            return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
```
[the remark this answers is not in the document]
Comment (ColdenCullen) [reply source/h1_connection.c]: I'll just not check the result, that's what Henso said to do goo.

File: source/h1_connection.c [hunk 12]
```diff
@@ -399,16 +410,24 @@ struct aws_http_stream *s_new_client_request_stream(const struct aws_http_reques
...
         goto error_scanning_headers;
     }
 
+    struct aws_byte_cursor method;
+    err |= aws_http_request_get_method(options->request, &method);
+    struct aws_byte_cursor uri;
+    err |= aws_http_request_get_path(options->request, &uri);
+    if (err) {
+        goto error_calculating_size;
+    }
```
[the remark this answers is not in the document]
Comment (ColdenCullen) [reply source/h1_connection.c]: Good call, I'll move this up.

File: source/h1_connection.c (hunk 3 above)
Comment (ColdenCullen) [reply source/h1_connection.c]: Will raise `AWS_IO_STREAM_READ_FAILED`

File: source/h1_connection.c [hunk 13]
```diff
@@ -410,20 +417,12 @@ struct aws_http_stream *s_new_client_request_stream(const struct aws_http_reques
      */
 
     size_t header_lines_len;
-    int err = s_stream_scan_outgoing_headers(stream, options->request, &header_lines_len);
+    err |= s_stream_scan_outgoing_headers(stream, options->request, &header_lines_len);
```
Comment (graebm) [inline source/h1_connection.c]: trivial: no reason for this to be `|=`

File: source/http.c [hunk 14]
```diff
@@ -147,6 +147,7 @@ static struct aws_hash_table s_method_str_to_enum;                         /* fo
 static struct aws_byte_cursor s_method_enum_to_str[AWS_HTTP_METHOD_COUNT]; /* for enum -> string lookup */
 
 static void s_methods_init(struct aws_allocator *alloc) {
+    s_method_enum_to_str[AWS_HTTP_METHOD_GET] = aws_http_method_get;
     s_method_enum_to_str[AWS_HTTP_METHOD_HEAD] = aws_byte_cursor_from_c_str("HEAD");
```
Comment (justinboswell) [inline source/http.c]: HEAD should be a constant along with the others

Comment (graebm) [inline source/http.c]: you can use your groovy new `aws_http_method_head`

**Pull request merged**